### DISCUSSION TITLE
ClusterBuilder EOF error when pushing the builder image

### DISCUSSION THREAD

**Discussion by semmet95** [buildpacks-community/kpack#1267] [Q&A]
Hey everyone.
I've been trying to use [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) to authenticate a ClusterBuilder to push builder images to an ECR repo and I keep running into the following error:
```
Post "https://<ecr-registry>/v2/kpack-builders/blobs/uploads/?from=paketobuildpacks%2Fbuild&mount=sha256%3A04519a37e0a828er368530248f886b2f85d7d4713e5a8a0ea1b419ter47663df&origin=index.docker.io": EOF
```
I couldn't find more details in `kpack-controller` logs either ☹️

![image](https://github.com/buildpacks-community/kpack/assets/18736972/46937a2b-5d5f-48e3-9461-70c03a169728)

I even tried using `docker-registry` type secret created as follows:
```
kubectl create secret docker-registry regcred --docker-server=<ecr-registry> --docker-username=AWS --docker-password=$(aws ecr get-login-password)
```
and mounted it on the serviceaccount but I still get the same error.

I have verified that the IAM role being used has all the necessary permissions and the target repo exists.
Has anyone else run into this same issue?
Any help would be appreciated 🙏🏻

**Comment by tomkennedy513** [comment]
Can you share how you are setting up the IAM roles to be used in kpack

**Reply by semmet95** [reply]
I'm following the instructions from AWS [docs](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html). I already had an IAM role with ECR permissions. I added the cluster's OIDC to its trusted entities and also made sure to explicitly mention the exact service account that would assume this role. Then I added the annotation `eks.amazonaws.com/role-arn: arn:aws:iam::111122223333:role/my-role` to the serviceaccount to let it assume the IAM role.

I have one question regarding the permissions. Do I need to add `ecr:GetAuthorizationToken` to the roles's actions? Currently I have added `ecr:Put*` for pushing permissions.

**Comment by semmet95** [comment]
Update: I found that kpack uses the corresponding worker node's role instead of the role associated to the attached serviceaccount via IRSA, to push the ClusterBuilder image.
Is this behaviour expected?

**Comment by semmet95** [comment]
To anyone stumbling upon this thread, it seems kpack doesn't support IRSA for ClusterBuilder, the image push permissions are determined by permissions the node `kpack-controller` is running on has, and not the role you attach to the serviceaccount via IRSA.

**Comment by alexbarbato** [comment]
FWIW - I think that this works now and uses the SA for `kpack-controller`

**Reply by chenbh** [reply]
I think this should've always worked (at least I'm looking at some of our internal EKS tests and they haven't changed in at least a year). The key is that the role must be attached to the `kpack-controller` service account in the `kpack` namespace.

The service accounts specified in the [Cluster]Builder will only be used for the secrets/dockerconfig-based auth, there's no mechanism to retrieve OIDC tokens from it

**Reply by tbarizien** [reply]
If someone stumbles upon this, we managed to solve the absence of mechanism to retrieve OIDC tokens by mounting an ECR authorization token into a secret attached to the service account and refreshing that token using the external secret operator (as described [here](https://external-secrets.io/latest/api/generator/ecr/) and in combination with some IRSA)
That way the controller can used a secret based auth and you don't need to give it extensive push rights to a bunch of ECRs
